onnxruntime/core/graph/graph_utils.cc

```diff
@@ -77,25 +77,128 @@ bool IsSingleInSingleOutNode(const Node& node) {
+                         [&removed_output_name](const NodeArg* input) {
+                           return input->Name() == removed_output_name;
+                         });
+
+        if (subgraph_node_also_consumes_nodearg_as_implicit_input != subgraph_node_implicit_inputs.cend()) {
+          UpdateImplicitInputNameInSubgraph(subgraph_node, removed_output_name, new_output_name);
+        }
+      }
+
+      // Need mutable input defs to be able to update the implicit input names
+      auto& input_args = subgraph_node.MutableInputDefs();
+
+      if (!input_args.empty()) {
+        int input_slot_index = -1;
+        for (const auto* input_arg : input_args) {
+          ++input_slot_index;
+          // if the input matches, replace the NodeArg with one using the new name
+          if (input_arg->Exists() && input_arg->Name() == removed_output_name) {
+            // sanity check there was no edge for this input. implicit inputs from outer scope do not have edges
+            ORT_ENFORCE(std::count_if(subgraph_node.InputEdgesBegin(), subgraph_node.InputEdgesEnd(),
+                                      [input_slot_index](const Node::EdgeEnd& entry) {
+                                        return entry.GetDstArgIndex() == input_slot_index;
+                                      }) == 0);
+
+            // Create a new NodeArg with the new name
+            input_args[input_slot_index] = &attr_subgraph_pair.second->GetOrCreateNodeArg(new_output_name,
+                                                                                          input_arg->TypeAsProto());
+          }
+        }
+      }
+    }
+  }
+}
+
 bool RemoveSingleInSingleOutNode(Graph& graph, Node& node) {
   if (!IsSingleInSingleOutNode(node)) {
     return false;
   }
+
   // Get input/output edges, nodes, and node args.
   const Node::EdgeEnd& input_edge = *node.InputEdgesBegin();
-  const NodeIndex input_edge_node = input_edge.GetNode().Index();
+  const Node& input_edge_node = input_edge.GetNode();
   const int input_edge_dst_arg = input_edge.GetSrcArgIndex();
   const Node::EdgeEnd& output_edge = *node.OutputEdgesBegin();
-  const NodeIndex output_edge_node = output_edge.GetNode().Index();
+  const Node& output_edge_node = output_edge.GetNode();
   const int output_edge_dst_arg = output_edge.GetDstArgIndex();
 
+  // check if the output from the node to remove is implicit input to the downstream node.
+  // if so, it's used in subgraph in that node and we have to check if it's valid to update that subgraph
+  bool is_implicit_input_to_subgraph = output_edge_dst_arg >= output_edge.GetNode().InputDefs().size();
+  if (is_implicit_input_to_subgraph) {
+    // the output we need to wire to the downstream node, and the output name from the node we want to remove
+    const auto& output_name = input_edge_node.OutputDefs()[output_edge.GetSrcArgIndex()]->Name();
+    const auto& removed_output_name = node.OutputDefs()[0]->Name();
```

Comment:
Shall we create a method that get the NodeArg name corresponding to an EdgeEnd? It seems quite useful in many cases. I think that could be part of the EdgeEnd actually (if not, we can create a helper method here).

Reply:
Sure. I'll add it to EdgeEnd most likely.

Reply:
Same issue here. EdgeEnd doesn't know if it's an input or output edge, so whether the src index or the dst index applies to the node instance it contains is unknown.

Reply:
Indeed... I think it's fine to leave this as is (or create a helper with a boolean in/out, but not sure it's worth it).